### ISSUE TITLE
MessageBodyWriter is called with genericType CompletionStage<T> when the resource method returns CompletionStage<T>

### ISSUE THREAD

**Issue by everett1992** [eclipse-ee4j/jersey#4640]
After executing a ResourceMethod with return type `CompletionStage<T>` the MessageBodyWriter is called with an object of type T, a type of T, but a genericType of `CompletionStage<T>`.

The spec is a little vague on CompletionStage, but I think the genericType should unwrap the CompletionStage and just be `T`.

> 3.3.3
>  Results in an entity body mapped from the class of the returned instance or of its type parameter T if the return type isCompletionStage<T>(see Section 8.2.2); if the class is an anonymous innerclass, its superclass is used instead. If the return value is notnulla 200 status code is used, anullreturn value results in a 204 status code.  
https://download.oracle.com/otn-pub/jcp/jaxrs-2_1-final-eval-spec/jaxrs-2_1-final-spec.pdf?AuthParam=1605738947_546c3c96f63881e7441d4b95f55d00e7

The current behavior is incongruous with, say, a ResourceMethod that returns Response - where MessageBodyWriter is called with the Response's entity, the type of it's entity, and the generic type of it's entity. e.g. a method that responds with `Resource.ok("test")` causes `MessageBodyWriter("test", String, String)`, not `("test", String, Response)`.

It looks like https://github.com/eclipse-ee4j/jersey/issues/3672 describes the same issue but the [PR](https://github.com/eclipse-ee4j/jersey/pull/3978) tries to fix it in jackson, when I think the fix should be be made in Jersey if my understanding of the spec is correct. I found the same issue using the [GsonMessageBodyProvider](https://github.com/immutables/immutables/blob/master/gson/src/org/immutables/gson/stream/GsonMessageBodyProvider.java) from [immutables](https://github.com/immutables/immutables) and it seems like a foot gun for any code using the genericType.

**Comment by everett1992** [comment]
I think this should be be false when `invocableResponseType` is assignable from `CompletionStage`, or `invocableResponseType` should be unwrapped to `T` if it is ``CompletionStage<T>`.

https://github.com/eclipse-ee4j/jersey/blob/master/core-server/src/main/java/org/glassfish/jersey/server/model/ResourceMethodInvoker.java#L304-L308

**Comment by jansupol** [comment]
This looks reasonable, but it changes the behaviour for message body providers. The `<CompletionStage>` is there for a long time, there would be a provider that expects `CompletionStage` which would have been broken by the changed behaviour.

There would be needed a configuration option to unwrap the `CompletionStage`.

**Comment by everett1992** [comment]
That makes sense, and I assume the current behavior would have to be the default? What about Jersey 3.0 or a future version of 2.?

**Comment by jansupol** [comment]
Yes, for Jersey 2.x the old behaviour should be the default. We already released Jersey 3.0.0, too.

**Comment by jansupol** [comment]
We currently go with [the following](https://github.com/eclipse-ee4j/jersey/pull/4643).

**Comment by everett1992** [comment]
Any documentation changes I can review or help with?

**Comment by everett1992** [comment]
This issue seems related https://github.com/eclipse-ee4j/jersey/issues/4463 and the root cause seems to be lack of detail in the JAX-RS spec

These are the only relevant references to  CompletionStage I could find and they are light on details, especially regarding exceptions.
https://github.com/eclipse-ee4j/jaxrs-api/blob/3c4941be798aba7986916b26e3f17e5b6c4ef85a/jaxrs-spec/src/main/asciidoc/chapters/asynchronous_processing/_server_api.adoc#completionstage
https://github.com/eclipse-ee4j/jaxrs-api/blob/9a89288dda19a66e2b72f04a14d993f1362b3549/jaxrs-spec/src/main/asciidoc/chapters/resources/_resource_method.adoc#return-type

**Comment by jansupol** [comment]
We will be able to change the default behaviour for Jakarta RESTful Web Services 3.1. implementation (possibly Jersey 3.1). It would be nice to mention it in the [3.x migration guide](https://github.com/eclipse-ee4j/jersey/blob/3.x/docs/src/main/docbook/migration.xml), then.
